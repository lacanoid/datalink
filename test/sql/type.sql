\pset null _null_

SET client_min_messages = warning;
SET search_path=public;

select dlvalue(null);
select dlvalue('');
select dlvalue('foo');
select dlvalue('http://www.ljudmila.org/');
select dlvalue('http://www.ljudmila.org/','URL','Example datalink');
select dlvalue('HTTP://WWW.ljudmila.org/A/B/c','URL');
select dlvalue('/tmp','FS','Example file datalink');
select dlvalue('file:///tmp','URL','Example file datalink #2');
select dlvalue('file:///tmp/a/b/c/d/../../x/y','URL','Example file datalink #3');

select dlvalue('favicon.ico',dlvalue('http://www.ljudmila.org/index.html'));
select dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced',dlvalue('/var/www/datalink/'),'Comment');
select dlurlpath(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced',dlvalue('/var/www/datalink/')));
select dlurlpathonly(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced',dlvalue('/var/www/datalink/')));

select dlvalue(null,null,'a comment');

select dlcomment(dlvalue('http://www.ljudmila.org/','URL','Example datalink'));
select dlurlcomplete(dlvalue('http://www.ljudmila.org/','URL','Example datalink'));
select dlurlcompleteonly(dlvalue('http://www.ljudmila.org/','URL','Example datalink'));

select dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0);
select dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1);
select dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0)::jsonb - 'b';
select dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1)::jsonb - 'b';

select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),0)::jsonb - 'b';
select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),0)::jsonb - 'b';
select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),1)::jsonb - 'b';
select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),1);

select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),0)::jsonb - 'b';
select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),0)::jsonb - 'b';
select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),1)::jsonb - 'b';
select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),1)::jsonb - 'b';

select dlnewcopy(dlvalue('http://www.ljudmila.org/'),0)::jsonb - 'b';
select (dlnewcopy(dlvalue('http://www.ljudmila.org'),1))::jsonb->>'b' is not null;

select dlpreviouscopy(dlvalue('http://www.ljudmila.org/'),0);
select (dlpreviouscopy(dlvalue('http://www.ljudmila.org'),1))::jsonb->>'b' is not null;

select dlpreviouscopy(dlvalue('/tmp/file1','FS'),0);
select key,case when key is distinct from 'b' then value end as value
  from jsonb_each(dlpreviouscopy(dlvalue('/tmp/file1','FS','Previous copy'),1)::jsonb);
select dlpreviouscopy(dlvalue('/tmp/file1#e5ed6a45-dc2f-42d2-a746-10c368677121','FS'),0);
select dlpreviouscopy(dlvalue('/tmp/file1#e5ed6a45-dc2f-42d2-a746-10c368677121','FS'),1);
select dlpreviouscopy('file:///tmp/file1#e5ed6a45-dc2f-42d2-a746-10c368677121',1);
select dlpreviouscopy('file:///tmp/file1#krneki',1);

select dlurlserver(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
select dlurlscheme(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));

select dlurlserver(dlvalue('HtTp://WwW.LjUDmILA.OrG/Foo/Bar/Baz#123'));
select dlurlscheme(dlvalue('HtTp://WwW.LjUDmILA.OrG/Foo/Bar/Baz#123'));

select dlurlpath(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
select dlurlpathonly(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));

select dlurlpath(dlvalue('file:///foo/bar/baz#123'));
select dlurlpathonly(dlvalue('file:///foo/bar/baz#123'));

select dllinktype(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
select dllinktype(dlvalue('/etc/issue','FS'));

select dlvalue('/etc/züöl');
select dlurlpath(dlvalue('/etc/züöl'));

select dlvalue('/etc/foo[1](2)#3');
select dlurlpath(dlvalue('/etc/foo[1](2)#3'));

select dlvalue('/etc/issue');
select dlvalue('/etc/issue','URL');
select dlvalue('/etc/issue','FS');
select dlvalue('/etc/issue','FILE');
select dlvalue('/etc/issue','foo');
select dllinktype(dlvalue('/etc/issue'));
select dllinktype(dlvalue('/etc/issue','FS'));
select dllinktype(dlvalue('/etc/issue','FILE'));
select dllinktype(dlvalue('/etc/issue','foo'));

select dlvalue('/foo/bar//baz/qux'); -- ERROR
select dlvalue('/foo/bar/../baz/qux'); -- ERROR
select dlvalue('/foo/bar/%/baz/qux'); -- ERROR

SET search_path=datalink;

select dl_lco(link_control=>'FILE');
select dl_lco(link_control=>'FILE',integrity=>'ALL',read_access=>'DB',write_access=>'BLOCKED');
select * from link_control_options(dl_lco(link_control=>'NO'));
select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'SELECTIVE'));
select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'ALL'));
select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'ALL',
					  read_access=>'DB',write_access=>'BLOCKED',
					  recovery=>'YES',on_unlink=>'DELETE'));
select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'ALL',
					  read_access=>'DB',write_access=>'BLOCKED',
					  recovery=>'YES',on_unlink=>'RESTORE'));

create table bar ( link datalink(123) );
begin;
create table foo ( link datalink(172) );
select * from datalink.columns where table_name in ('foo','bar');
abort;

select dlvalue('robots.txt',dlvalue('http://localhost/index.html',null,'foo'));
select dlvalue('robots.txt',dlvalue('http://localhost/index.html',null,'foo'),'bar');

