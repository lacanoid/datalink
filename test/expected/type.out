\pset null _null_
SET client_min_messages = warning;
SET search_path=public;
select dlvalue(null);
 dlvalue 
---------
 _null_
(1 row)

select dlvalue('');
 dlvalue 
---------
 _null_
(1 row)

select dlvalue('foo');
ERROR:  value for domain datalink.url violates check constraint "url_check"
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 41 at assignment
select dlvalue('http://www.ljudmila.org/');
              dlvalue              
-----------------------------------
 {"a": "http://www.ljudmila.org/"}
(1 row)

select dlvalue('http://www.ljudmila.org/','URL','Example datalink');
                          dlvalue                           
------------------------------------------------------------
 {"a": "http://www.ljudmila.org/", "c": "Example datalink"}
(1 row)

select dlvalue('HTTP://WWW.ljudmila.org/A/B/c','URL');
                dlvalue                 
----------------------------------------
 {"a": "http://www.ljudmila.org/A/B/c"}
(1 row)

select dlvalue('/tmp','FS','Example file datalink');
                      dlvalue                       
----------------------------------------------------
 {"a": "file:///tmp", "c": "Example file datalink"}
(1 row)

select dlvalue('file:///tmp','URL','Example file datalink #2');
                        dlvalue                        
-------------------------------------------------------
 {"a": "file:///tmp", "c": "Example file datalink #2"}
(1 row)

select dlvalue('file:///tmp/a/b/c/d/../../x/y','URL','Example file datalink #3');
                            dlvalue                            
---------------------------------------------------------------
 {"a": "file:///tmp/a/b/x/y", "c": "Example file datalink #3"}
(1 row)

select dlvalue('favicon.ico',dlvalue('http://www.ljudmila.org/index.html'));
                   dlvalue                    
----------------------------------------------
 {"a": "http://www.ljudmila.org/favicon.ico"}
(1 row)

select dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced',dlvalue('/var/www/datalink/'),'Comment');
                                             dlvalue                                              
--------------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced", "c": "Comment"}
(1 row)

select dlurlpath(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced',dlvalue('/var/www/datalink/')));
                            dlurlpath                             
------------------------------------------------------------------
 /var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced
(1 row)

select dlurlpathonly(dlvalue('test3.txt#11111111-2222-3333-4444-abecedabeced',dlvalue('/var/www/datalink/')));
        dlurlpathonly        
-----------------------------
 /var/www/datalink/test3.txt
(1 row)

select dlvalue(null,null,'a comment');
      dlvalue       
--------------------
 {"c": "a comment"}
(1 row)

select dlcomment(dlvalue('http://www.ljudmila.org/','URL','Example datalink'));
    dlcomment     
------------------
 Example datalink
(1 row)

select dlurlcomplete(dlvalue('http://www.ljudmila.org/','URL','Example datalink'));
      dlurlcomplete       
--------------------------
 http://www.ljudmila.org/
(1 row)

select dlurlcompleteonly(dlvalue('http://www.ljudmila.org/','URL','Example datalink'));
    dlurlcompleteonly     
--------------------------
 http://www.ljudmila.org/
(1 row)

select dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0);
                                       dlpreviouscopy                                       
--------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced", "k": "p"}
(1 row)

select dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1);
                                           dlpreviouscopy                                           
----------------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "b": "11111111-2222-3333-4444-abecedabeced", "k": "p"}
(1 row)

select dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0)::jsonb - 'b';
                       ?column?                        
-------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "n"}
(1 row)

select dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1)::jsonb - 'b';
                                              ?column?                                              
----------------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "n", "o": "11111111-2222-3333-4444-abecedabeced"}
(1 row)

select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),0)::jsonb - 'b';
                       ?column?                        
-------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "p"}
(1 row)

select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),0)::jsonb - 'b';
                                              ?column?                                              
----------------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "p", "o": "11111111-2222-3333-4444-abecedabeced"}
(1 row)

select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),1)::jsonb - 'b';
                       ?column?                        
-------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "p"}
(1 row)

select dlpreviouscopy(dlnewcopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),1);
                                           dlpreviouscopy                                           
----------------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "b": "11111111-2222-3333-4444-abecedabeced", "k": "p"}
(1 row)

select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),0)::jsonb - 'b';
                       ?column?                        
-------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "n"}
(1 row)

select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),0)::jsonb - 'b';
                       ?column?                        
-------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "n"}
(1 row)

select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),0),1)::jsonb - 'b';
                                              ?column?                                              
----------------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "n", "o": "11111111-2222-3333-4444-abecedabeced"}
(1 row)

select dlnewcopy(dlpreviouscopy(dlvalue('/var/www/datalink/test3.txt#11111111-2222-3333-4444-abecedabeced'),1),1)::jsonb - 'b';
                                              ?column?                                              
----------------------------------------------------------------------------------------------------
 {"a": "file:///var/www/datalink/test3.txt", "k": "n", "o": "11111111-2222-3333-4444-abecedabeced"}
(1 row)

select dlnewcopy(dlvalue('http://www.ljudmila.org/'),0)::jsonb - 'b';
                  ?column?                   
---------------------------------------------
 {"a": "http://www.ljudmila.org/", "k": "n"}
(1 row)

select (dlnewcopy(dlvalue('http://www.ljudmila.org'),1))::jsonb->>'b' is not null;
 ?column? 
----------
 t
(1 row)

select dlpreviouscopy(dlvalue('http://www.ljudmila.org/'),0);
               dlpreviouscopy                
---------------------------------------------
 {"a": "http://www.ljudmila.org/", "k": "p"}
(1 row)

select (dlpreviouscopy(dlvalue('http://www.ljudmila.org'),1))::jsonb->>'b' is not null;
 ?column? 
----------
 t
(1 row)

select dlpreviouscopy(dlvalue('/tmp/file1','FS'),0);
            dlpreviouscopy            
--------------------------------------
 {"a": "file:///tmp/file1", "k": "p"}
(1 row)

select key,case when key is distinct from 'b' then value end as value
  from jsonb_each(dlpreviouscopy(dlvalue('/tmp/file1','FS','Previous copy'),1)::jsonb);
 key |        value        
-----+---------------------
 a   | "file:///tmp/file1"
 b   | _null_
 c   | "Previous copy"
 k   | "p"
(4 rows)

select dlpreviouscopy(dlvalue('/tmp/file1#e5ed6a45-dc2f-42d2-a746-10c368677121','FS'),0);
                              dlpreviouscopy                               
---------------------------------------------------------------------------
 {"a": "file:///tmp/file1#e5ed6a45-dc2f-42d2-a746-10c368677121", "k": "p"}
(1 row)

select dlpreviouscopy(dlvalue('/tmp/file1#e5ed6a45-dc2f-42d2-a746-10c368677121','FS'),1);
                                  dlpreviouscopy                                   
-----------------------------------------------------------------------------------
 {"a": "file:///tmp/file1", "b": "e5ed6a45-dc2f-42d2-a746-10c368677121", "k": "p"}
(1 row)

select dlpreviouscopy('file:///tmp/file1#e5ed6a45-dc2f-42d2-a746-10c368677121',1);
                                  dlpreviouscopy                                   
-----------------------------------------------------------------------------------
 {"a": "file:///tmp/file1", "b": "e5ed6a45-dc2f-42d2-a746-10c368677121", "k": "p"}
(1 row)

select dlpreviouscopy('file:///tmp/file1#krneki',1);
ERROR:  DATALINK EXCEPTION - invalid write token
DETAIL:  invalid input syntax for type uuid: "krneki"
CONTEXT:  PL/pgSQL function dlpreviouscopy(datalink,integer) line 18 at RAISE
select dlurlserver(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
   dlurlserver    
------------------
 www.ljudmila.org
(1 row)

select dlurlscheme(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
 dlurlscheme 
-------------
 HTTP
(1 row)

select dlurlpath(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
    dlurlpath     
------------------
 /foo/bar/baz#123
(1 row)

select dlurlpathonly(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
 dlurlpathonly 
---------------
 /foo/bar/baz
(1 row)

select dlurlpath(dlvalue('file:///foo/bar/baz#123'));
    dlurlpath     
------------------
 /foo/bar/baz#123
(1 row)

select dlurlpathonly(dlvalue('file:///foo/bar/baz#123'));
 dlurlpathonly 
---------------
 /foo/bar/baz
(1 row)

select dllinktype(dlvalue('http://www.ljudmila.org/foo/bar/baz#123'));
 dllinktype 
------------
 URL
(1 row)

select dllinktype(dlvalue('/etc/issue','FS'));
 dllinktype 
------------
 FS
(1 row)

select dlvalue('/etc/züöl');
               dlvalue               
-------------------------------------
 {"a": "file:///etc/z%C3%BC%C3%B6l"}
(1 row)

select dlurlpath(dlvalue('/etc/züöl'));
 dlurlpath 
-----------
 /etc/züöl
(1 row)

select dlvalue('/etc/foo[1](2)#3');
                 dlvalue                  
------------------------------------------
 {"a": "file:///etc/foo%5B1%5D%282%29#3"}
(1 row)

select dlurlpath(dlvalue('/etc/foo[1](2)#3'));
    dlurlpath     
------------------
 /etc/foo[1](2)#3
(1 row)

select dlvalue('/etc/issue');
          dlvalue           
----------------------------
 {"a": "file:///etc/issue"}
(1 row)

select dlvalue('/etc/issue','URL');
ERROR:  value for domain datalink.url violates check constraint "url_check"
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 41 at assignment
select dlvalue('/etc/issue','FS');
          dlvalue           
----------------------------
 {"a": "file:///etc/issue"}
(1 row)

select dlvalue('/etc/issue','FILE');
ERROR:  DATALINK EXCEPTION - nonexistent directory
DETAIL:  directory "FILE" does not exist
HINT:  perhaps you need to add it to datalink.directory
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 27 at RAISE
select dlvalue('/etc/issue','foo');
ERROR:  DATALINK EXCEPTION - nonexistent directory
DETAIL:  directory "foo" does not exist
HINT:  perhaps you need to add it to datalink.directory
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 27 at RAISE
select dllinktype(dlvalue('/etc/issue'));
 dllinktype 
------------
 FS
(1 row)

select dllinktype(dlvalue('/etc/issue','FS'));
 dllinktype 
------------
 FS
(1 row)

select dllinktype(dlvalue('/etc/issue','FILE'));
ERROR:  DATALINK EXCEPTION - nonexistent directory
DETAIL:  directory "FILE" does not exist
HINT:  perhaps you need to add it to datalink.directory
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 27 at RAISE
select dllinktype(dlvalue('/etc/issue','foo'));
ERROR:  DATALINK EXCEPTION - nonexistent directory
DETAIL:  directory "foo" does not exist
HINT:  perhaps you need to add it to datalink.directory
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 27 at RAISE
select dlvalue('/foo/bar//baz/qux'); -- ERROR
ERROR:  value for domain datalink.file_path violates check constraint "file_path_noserver"
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 35 at assignment
select dlvalue('/foo/bar/../baz/qux'); -- ERROR
ERROR:  value for domain datalink.file_path violates check constraint "file_path_noparent"
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 35 at assignment
select dlvalue('/foo/bar/%/baz/qux'); -- ERROR
ERROR:  value for domain datalink.file_path violates check constraint "file_path_chars"
CONTEXT:  PL/pgSQL function dlvalue(text,datalink.dl_linktype,text) line 35 at assignment
SET search_path=datalink;
select dl_lco(link_control=>'FILE');
 dl_lco 
--------
      0
(1 row)

select dl_lco(link_control=>'FILE',integrity=>'ALL',read_access=>'DB',write_access=>'BLOCKED');
 dl_lco 
--------
     52
(1 row)

select * from link_control_options(dl_lco(link_control=>'NO'));
 lco | link_control | integrity | read_access | write_access | recovery | on_unlink 
-----+--------------+-----------+-------------+--------------+----------+-----------
   0 | NO           | NONE      | FS          | FS           | NO       | NONE
(1 row)

select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'SELECTIVE'));
 lco | link_control | integrity | read_access | write_access | recovery | on_unlink 
-----+--------------+-----------+-------------+--------------+----------+-----------
   1 | FILE         | SELECTIVE | FS          | FS           | NO       | NONE
(1 row)

select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'ALL'));
 lco | link_control | integrity | read_access | write_access | recovery | on_unlink 
-----+--------------+-----------+-------------+--------------+----------+-----------
   2 | FILE         | ALL       | FS          | FS           | NO       | NONE
(1 row)

select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'ALL',
					  read_access=>'DB',write_access=>'BLOCKED',
					  recovery=>'YES',on_unlink=>'DELETE'));
 lco | link_control | integrity | read_access | write_access | recovery | on_unlink 
-----+--------------+-----------+-------------+--------------+----------+-----------
 352 | FILE         | ALL       | DB          | BLOCKED      | YES      | DELETE
(1 row)

select * from link_control_options(dl_lco(link_control=>'FILE',integrity=>'ALL',
					  read_access=>'DB',write_access=>'BLOCKED',
					  recovery=>'YES',on_unlink=>'RESTORE'));
 lco | link_control | integrity | read_access | write_access | recovery | on_unlink 
-----+--------------+-----------+-------------+--------------+----------+-----------
 152 | FILE         | ALL       | DB          | BLOCKED      | YES      | RESTORE
(1 row)

create table bar ( link datalink(123) );
ERROR:  DATALINK EXCEPTION
DETAIL:  Invalid link control options
HINT:  see table datalink.link_control_options for valid link control options
CONTEXT:  PL/pgSQL function dl_trigger_event() line 43 at RAISE
begin;
create table foo ( link datalink(172) );
select * from datalink.columns where table_name in ('foo','bar');
 table_name | column_name | link_control | integrity | read_access | write_access | recovery | on_unlink 
------------+-------------+--------------+-----------+-------------+--------------+----------+-----------
 foo        | link        | FILE         | ALL       | DB          | ADMIN        | YES      | RESTORE
(1 row)

abort;
select dlvalue('robots.txt',dlvalue('http://localhost/index.html',null,'foo'));
                     dlvalue                      
--------------------------------------------------
 {"a": "http://localhost/robots.txt", "c": "foo"}
(1 row)

select dlvalue('robots.txt',dlvalue('http://localhost/index.html',null,'foo'),'bar');
                     dlvalue                      
--------------------------------------------------
 {"a": "http://localhost/robots.txt", "c": "bar"}
(1 row)

